src/gui/widgets/grid.cpp

```diff
@@ -89,52 +89,59 @@ void grid::set_child(widget* widget,
 	// copy data
 	cell.set_flags(flags);
 	cell.set_border_size(border_size);
-	cell.set_widget(widget);
-	if(cell.get_widget()) {
-		// make sure the new child is valid before deferring
-		cell.get_widget()->set_parent(this);
+	cell.set_widget(std::move(widget));
+
+	// make sure the new child is valid before deferring
+	if(gui2::widget* w = cell.get_widget()) {
+		w->set_parent(this);
 	}
 }
 
 std::unique_ptr<widget> grid::swap_child(const std::string& id,
-						   widget* w,
-						   const bool recurse,
-						   widget* new_parent)
+		std::unique_ptr<widget> w,
+		const bool recurse,
+		widget* new_parent)
 {
 	assert(w);
 
-	for(auto & child : children_)
-	{
-		if(child.id() != id) {
+	std::vector<grid*> child_grids{};
 
+	for(auto& child : children_) {
+		if(child.id() != id) {
 			if(recurse) {
-				// decent in the nested grids.
-				grid* g = dynamic_cast<grid*>(child.get_widget());
-				if(g) {
-
-					std::unique_ptr<widget> old = g->swap_child(id, w, true);
-					if(old) {
-						return old;
-					}
+				// Descend into nested grids. We don't handle recursion here for two reasons:
```

Comment:
This lengthy comment seems to suggest that the new code is not equivalent to the old code at all? But I have no idea what implications that would have.

Reply:
It is, I just defer handling child grids until we've checked all widgets... though I'm just realizing the logic still doesn't work, since there may be more than one child grid...

Reply:
What do you mean, "it is"? How exactly is it equivalent?

Reply:
It still looks at every widget and potentially sub-grids. I'm not sure why you're saying it's not equivalent at all.

Reply:
It looks like things happen in an entirely different order, and I have no idea what consequences that might have. It might give a different result in some cases, right?

Reply:
I've looked through all the places that call `swap_child` in this diff, and while I'm not 100% sure that's all calls in the entire game, it looks like this is the only place that relies on it returning `nullptr` if nothing was swapped. So… couldn't the owning problem be solved by instead returning the passed-in widget if nothing was swapped? Something like this:

```cpp
widget* test = w.get();
std::unique_ptr<widget> old = g->swap_child(id, w, true)
if(test != old) return old;
```

Reply:
Well, it's just looking for the widget matching the given ID, so it shouldn't matter in what order it happens.

Reply:
There could be more than one widget with the ID. There's literally nothing to enforce that IDs are unique. In fact, they're guaranteed _not_ to be unique if you're using anything like a listbox… though I don't know how relevant that is in this specific case.

Although neither the old nor the new could be called a sensible way of handling duplicate IDs, it certainly does change how they'd be handled.

Reply:
> ```c++
> widget* test = w.get();
> std::unique_ptr<widget> old = g->swap_child(id, w, true)
> if(test != old) return old;
> ```
I like this. Thanks!